tests/Microsoft.Identity.Test.Unit/AuthExtension/AuthenticationOperationTests.cs

```diff
@@ -0,0 +1,133 @@
+                    .WithTenantId(TestConstants.Utid)
+                    .WithAuthenticationExtension(authExtension)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                var expectedAt = "header.payload.signature" + "AccessTokenModifier";
+                Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+                Assert.IsFalse(result.AdditionalResponseParameters.Any());
+                Assert.AreEqual(expectedAt, result.AccessToken);
+
+                //Verify that the original AT token is cached and the extension is reused
+                result = await app.AcquireTokenForClient(TestConstants.s_scope.ToArray())
+                    .WithTenantId(TestConstants.Utid)
+                    .WithAuthenticationExtension(authExtension)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+                Assert.IsTrue(result.AdditionalResponseParameters == null);
+                Assert.AreEqual(expectedAt, result.AccessToken);
+            }
+        }
+
+        [TestMethod]
+        public async Task AuthenticationOperationWithCachingTest_Async()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                ConfidentialClientApplication app =
+                    ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                              .WithClientSecret(TestConstants.ClientSecret)
+                                                              .WithHttpManager(httpManager)
+                                                              .WithExperimentalFeatures(true)
+                                                              .BuildConcrete();
+
+                HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, new Uri(ProtectedUrl));
+
+                Dictionary<string, string> expectedRequestHeaders = new Dictionary<string, string>
+                {
+                    { "key1", "value1" }
+                };
+                
+                httpManager.AddInstanceDiscoveryMockHandler();
+                httpManager.AddMockHandlerSuccessfulClientCredentialTokenResponseWithAdditionalParamsMessage(tokenType: "someAccessTokenType", expectedRequestHeaders: expectedRequestHeaders);
+
+                MsalAuthenticationExtension authExtension = new MsalAuthenticationExtension()
+                {
+                    OnBeforeTokenRequestHandler = async (data) =>
+                    {
+                        data.Headers.Add("key1", "value1");
+                        await Task.CompletedTask.ConfigureAwait(false);
+                    },
+
+                    AuthenticationOperation = new MsalTestAuthenticationOperation(),
```

Comment:
can we test for, 

AdditionalCacheParameters = null, // Testing null AdditionalCacheParameters
OnBeforeTokenRequestHandler = null // Testing null OnBeforeTokenRequestHandler